### PR TITLE
Leqiao/add static analysis

### PR DESCRIPTION
**Pull request by leqiao-1**
add static analysis

### CODE REVIEW

File: .pipelines/OneBranch.Official.yml
Comment (wenbingl) [inline .pipelines/OneBranch.Official.yml:92]: can this hard code path be inferred from vswherepath?

**Review by wenbingl — APPROVED** [review]

File: .pipelines/OneBranch.Official.yml
Comment (snnn) [inline .pipelines/OneBranch.Official.yml:94]: Here you should exclude  $(REPOROOT)\windows_out instead of $(Build.BinariesDirectory)

File: .pipelines/OneBranch.Official.yml
Comment (snnn) [inline .pipelines/OneBranch.Official.yml:89]: Do not let this task override the files in `$(REPOROOT)\out`. Please consider to use a different folder. Or you may upload the built python packages first.

File: .pipelines/OneBranch.Official.yml
Comment (snnn) [inline .pipelines/OneBranch.Official.yml:98]: You might need to append a "Guardian Break" task to detect if PREfast outputs any warnings.

Comment (wenbingl) [reply .pipelines/OneBranch.Official.yml:98]: There are already hundred of warnings caused by external 3rd party libraries. Won't expect it will be fixed.
Does PREfast only output warnings only, or something severer will be outputted as error?

Comment (snnn) [reply .pipelines/OneBranch.Official.yml:98]: _deps are in your build folder. See my comments above. Yes, we should exclude them.

Comment (snnn) [reply .pipelines/OneBranch.Official.yml:98]: This task output warnings only.  If you run it locally without this ADO task, warnings are errors

File: .pipelines/OneBranch.Official.yml
Comment (snnn) [reply .pipelines/OneBranch.Official.yml:94]: @wenbingl FYI.

Comment (leqiao-1) [reply .pipelines/OneBranch.Official.yml]: ```$(REPOROOT)\windows_out``` was excluede

File: .pipelines/OneBranch.Official.yml
Comment (leqiao-1) [reply .pipelines/OneBranch.Official.yml:99]: This PR doesn't modify files in ```$(REPOROOT)\out```. The output for static analysis is saved in ```windows_out``` folder.

File: .pipelines/OneBranch.Official.yml
Comment (leqiao-1) [reply .pipelines/OneBranch.Official.yml:113]: Guardian Break task was added.

Comment (snnn) [reply .pipelines/OneBranch.Official.yml:113]: But the static analysis task found 25 warnings and this job didn't fail. So maybe something else is also needed.

Comment (snnn) [reply .pipelines/OneBranch.Official.yml:113]: Try SdtReport task

Comment (snnn) [reply .pipelines/OneBranch.Official.yml:113]: So the "Guardian Break" task found 12 warnings but I don't know why they are skipped.

Guardian is searching for results that meet the given criteria to break the build.
  Results Query Summary:
    Tool Filters (Include): apiscan, armory, bandit, binskim, codesign, credscan, csrf, detekt, eslint, flawfinder, FortifySca, fxcop, gosec:Error, infersharp:Error, moderncop, policheck, psscriptanalyzer, roslynanalyzers, prefast:Warning, semmle, spotbugs, tslint, terrascan, templateanalyzer
    Baselines: default
    Suppression Sets: default
    Policy: Microsoft
  Found no breaking results.
  Active results: 0
  Skipped results: 12
    Baselined results: 0
    Suppressed results: 0
    Results excluded by tool filters: 0
    Results below minimum severity: 12
    Results classified as Pass: 0

Comment (snnn) [reply .pipelines/OneBranch.Official.yml:113]: Anyway, at least the analyzer is working now. We can merge this PR first and improve it later.

**Review by snnn — APPROVED** [review]

File: .pipelines/OneBranch.Official.yml
Comment (snnn) [inline .pipelines/OneBranch.Official.yml:103]: You will need to copy the ruleset file from ort repo to this repo, and use it. 
https://github.com/microsoft/onnxruntime/blob/master/cmake/Sdl.ruleset

By default this task enables far less checks.

Comment (leqiao-1) [reply .pipelines/OneBranch.Official.yml:103]: ruleset file added

Comment (snnn) [reply .pipelines/OneBranch.Official.yml:103]: Build log shows:
```
/p:GuardianPREfastRuleSet="C:\__w\_gdn\packages\nuget\Microsoft.Security.CodeAnalysis.PREfast.Cli.1.3.0\tools\Sdl.Recommended.Warning.ruleset"
```
Still not the one we want.

File: .pipelines/OneBranch.PullRequest.yml
Comment (snnn) [inline .pipelines/OneBranch.PullRequest.yml:99]: I think you need to specify the ruleset file here.

**Comment by snnn** [comment]
Please add the following line to this repo's CMakeLists.txt

https://github.com/microsoft/onnxruntime/blob/master/cmake/CMakeLists.txt#L584

**Comment by wenbingl** [comment]
Thanks, @leqiao-1 , @snnn , please go ahead and merge this PR and make some improvement if needed in the other PRs.
So, the PR #274 will be verified after this PR will be merge.

**Comment by leqiao-1** [comment]
> Thanks, @leqiao-1 , @snnn , please go ahead and merge this PR and make some improvement if needed in the other PRs. So, the PR #274 will be verified after this PR will be merge.

The review was dismissed when I updating this PR. We can merge this PR fisrt, and I will create a new PR for other modification. @wenbingl @snnn

**Review by wenbingl — APPROVED** [review]

**Pull request merged**
